api/filrewardsd/service/service_test.go

```diff
@@ -0,0 +1,498 @@
+	t.Parallel()
+	c, cleanup := requireSetup(t, ctx)
+	defer cleanup()
+	requireProcessedEvent(t, ctx, c, "user1", analyticspb.Event_EVENT_BILLING_SETUP)
+	time.Sleep(time.Millisecond * 500)
+	requireProcessedEvent(t, ctx, c, "user1", analyticspb.Event_EVENT_BUCKET_CREATED)
+	time.Sleep(time.Millisecond * 500)
+	requireProcessedEvent(t, ctx, c, "user1", analyticspb.Event_EVENT_BUCKET_ARCHIVE_CREATED)
+	res, err := c.List(ctx, &pb.ListRequest{Ascending: false})
+	require.NoError(t, err)
+	require.Len(t, res.RewardRecords, 3)
+	requireOrder(t, res.RewardRecords, false)
+}
+
+func TestListAscending(t *testing.T) {
+	t.Parallel()
+	c, cleanup := requireSetup(t, ctx)
+	defer cleanup()
+	requireProcessedEvent(t, ctx, c, "user1", analyticspb.Event_EVENT_BILLING_SETUP)
+	time.Sleep(time.Millisecond * 500)
+	requireProcessedEvent(t, ctx, c, "user1", analyticspb.Event_EVENT_BUCKET_CREATED)
+	time.Sleep(time.Millisecond * 500)
+	requireProcessedEvent(t, ctx, c, "user1", analyticspb.Event_EVENT_BUCKET_ARCHIVE_CREATED)
+	res, err := c.List(ctx, &pb.ListRequest{Ascending: true})
+	require.NoError(t, err)
+	require.Len(t, res.RewardRecords, 3)
+	requireOrder(t, res.RewardRecords, true)
+}
+
+func TestListPaging(t *testing.T) {
+	t.Parallel()
+	c, cleanup := requireSetup(t, ctx)
+	defer cleanup()
+	requireProcessedEvent(t, ctx, c, "user1", analyticspb.Event_EVENT_BILLING_SETUP)
+	time.Sleep(time.Millisecond * 500)
+	requireProcessedEvent(t, ctx, c, "user1", analyticspb.Event_EVENT_BUCKET_CREATED)
+	time.Sleep(time.Millisecond * 500)
+	requireProcessedEvent(t, ctx, c, "user1", analyticspb.Event_EVENT_BUCKET_ARCHIVE_CREATED)
+	time.Sleep(time.Millisecond * 500)
+	requireProcessedEvent(t, ctx, c, "user2", analyticspb.Event_EVENT_BILLING_SETUP)
+	time.Sleep(time.Millisecond * 500)
+	requireProcessedEvent(t, ctx, c, "user2", analyticspb.Event_EVENT_BUCKET_CREATED)
+	time.Sleep(time.Millisecond * 500)
+	requireProcessedEvent(t, ctx, c, "user2", analyticspb.Event_EVENT_BUCKET_ARCHIVE_CREATED)
+	time.Sleep(time.Millisecond * 500)
+	requireProcessedEvent(t, ctx, c, "user3", analyticspb.Event_EVENT_BILLING_SETUP)
+	time.Sleep(time.Millisecond * 500)
+	requireProcessedEvent(t, ctx, c, "user3", analyticspb.Event_EVENT_BUCKET_CREATED)
+	numPages := 0
+	more := true
+	var startAtToken *timestamppb.Timestamp
+	for more {
+		req := &pb.ListRequest{Limit: 3}
+		if startAtToken != nil {
+			req.StartAt = startAtToken
+		}
+		res, err := c.List(ctx, req)
+		require.NoError(t, err)
+		numPages++
+		if numPages < 3 {
+			require.True(t, res.More)
+			require.NotNil(t, res.MoreStartAt)
+			require.Len(t, res.RewardRecords, 3)
+		}
+		if numPages == 3 {
+			require.False(t, res.More)
+			require.Nil(t, res.MoreStartAt)
+			require.Len(t, res.RewardRecords, 2)
+		}
+		startAtToken = res.MoreStartAt
+		more = res.More
+	}
+	require.Equal(t, 3, numPages)
+}
+
+func requireSetup(t *testing.T, ctx context.Context) (pb.FilRewardsServiceClient, func()) {
+	listener := bufconn.Listen(bufSize)
```

Comment:
This is the listener you can use for testing that I described in the previous comment.